File: app/src/main/java/net/cynreub/subly/domain/usecase/GetSubscriptionStatsUseCase.kt
package net.cynreub.subly.domain.usecase

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import net.cynreub.subly.domain.model.BillingFrequency
import net.cynreub.subly.domain.repository.SubscriptionRepository
import java.util.UUID
import javax.inject.Inject

data class SubscriptionStats(
    val totalMonthly: Double,
    val totalYearly: Double,
    val activeCount: Int,
    val categoryBreakdown: Map<UUID, Double> // categoryId → monthly spend equivalent
)

class GetSubscriptionStatsUseCase @Inject constructor(
    private val subscriptionRepository: SubscriptionRepository
) {
    operator fun invoke(): Flow<SubscriptionStats> {
        return subscriptionRepository.getActiveSubscriptions()
            .map { subscriptions ->
                val totalMonthly = subscriptions.sumOf { it.toMonthlyAmount() }

                val categoryBreakdown = subscriptions
                    .groupBy { it.categoryId }
                    .mapValues { (_, subs) -> subs.sumOf { it.toMonthlyAmount() } }

                SubscriptionStats(
                    totalMonthly = totalMonthly,
                    totalYearly = totalMonthly * 12,
                    activeCount = subscriptions.size,
                    categoryBreakdown = categoryBreakdown
                )
            }
    }

    private fun net.cynreub.subly.domain.model.Subscription.toMonthlyAmount(): Double =
        when (frequency) {
            BillingFrequency.WEEKLY -> amount * 4
            BillingFrequency.MONTHLY -> amount
            BillingFrequency.QUARTERLY -> amount / 3
            BillingFrequency.SEMI_ANNUAL -> amount / 6
            BillingFrequency.ANNUAL -> amount / 12
            BillingFrequency.CUSTOM -> amount
        }
}

File: subly_ios/Subly/Subly/Domain/UseCases/GetSubscriptionStatsUseCase.swift
import Foundation
import Combine

struct SubscriptionStats {
    let totalMonthly: Double
    let totalYearly: Double
    let activeCount: Int
    let categoryBreakdown: [UUID: Double] // categoryId → monthly spend equivalent
}

final class GetSubscriptionStatsUseCase {
    private let subscriptionRepository: SubscriptionRepository

    init(subscriptionRepository: SubscriptionRepository) {
        self.subscriptionRepository = subscriptionRepository
    }

    func execute() -> AnyPublisher<SubscriptionStats, Never> {
        subscriptionRepository.getActiveSubscriptions()
            .map { subscriptions in
                let totalMonthly = subscriptions.reduce(0.0) { $0 + $1.monthlyAmount }

                let categoryBreakdown = Dictionary(
                    grouping: subscriptions,
                    by: { $0.categoryId }
                ).mapValues { subs in
                    subs.reduce(0.0) { $0 + $1.monthlyAmount }
                }

                return SubscriptionStats(
                    totalMonthly: totalMonthly,
                    totalYearly: totalMonthly * 12,
                    activeCount: subscriptions.count,
                    categoryBreakdown: categoryBreakdown
                )
            }
            .eraseToAnyPublisher()
    }
}
